### PR TITLE
[gatsby-source-filesystem] First parse url and then path to retrieve extension

### PR DESCRIPTION
**Pull request by oorestisime**
I was working on  a plugin that uses the createRemoteFileNode and my urls had the following form (comes from instagram api)
`https://scontent.xx.fbcdn.net/v/t51.2885-15/42078503_294439751160571_1602896118583132160_n.jpg?_nc_cat=101&oh=e30490a47409051c45dc3daacf616bc0&oe=5C5EA8EB`

It was creating the following file nodes
```
 internal: 
   { contentDigest: '961029972dbcc230aa2d81a006e48a82',
     type: 'File',
     mediaType: 'application/octet-stream',
     description: 'File "https://scontent.xx.fbcdn.net/v/t51.2885-15/40837333_345458009525698_761616599160455168_n.jpg?_nc_cat=105&oh=37800f1dc756dcd31f00482fd1a789eb&oe=5C5216F7"',
     owner: 'gatsby-source-filesystem' },
  sourceInstanceName: '__PROGRAMATTIC__',
  absolutePath: '/home/grok/personal/oasome/.cache/gatsby-source-filesystem/4899d60e50c2d4af9c4da15f2792a9dd.jpg?_nc_cat=105&oh=37800f1dc756dcd31f00482fd1a789eb&oe=5C5216F7',
  relativePath: '.cache/gatsby-source-filesystem/4899d60e50c2d4af9c4da15f2792a9dd.jpg?_nc_cat=105&oh=37800f1dc756dcd31f00482fd1a789eb&oe=5C5216F7',
  extension: 'jpg?_nc_cat=105&oh=37800f1dc756dcd31f00482fd1a789eb&oe=5c5216f7',
```

And the extension was basically broken.

This lead to weird bugs in gatsby image.  I.e querying for localfile like:

```
localFile {
          childImageSharp {
            fixed(width: 150, height:150, quality: 100) {
              src
              srcSet
              srcWebp
            }
          }
        }
```
was giving 
```
"localFile": {
     "childImageSharp": null
}
```

Haven't really checked why with that extension this breaks (maybe something to do with `&`), i can dig in more if you want, but with that fix everything works again perfectly (plus i think its normal to first parse the url and then the path)

### CODE REVIEW

**Comment by pieh** [comment]
> Haven't really checked why with that extension this breaks (maybe something to do with `&`), i can dig in more if you want, but with that fix everything works again perfectly (plus i think its normal to first parse the url and then the path)

We test extensions in `gatsby-transformer-sharp` and this weird extension isn't in allowed map: packages/gatsby/src/bootstrap/load-plugins/index.js

File: packages/gatsby-source-filesystem/src/create-remote-file-node.js [hunk 1]
```diff
@@ -195,7 +196,7 @@ async function processRemoteNode({
 
   // Create the temp and permanent file names for the url.
   const digest = createHash(url)
-  const ext = path.parse(url).ext
+  const ext = path.parse(Url.parse(url).pathname).ext
```
Comment (DSchau) [inline packages/gatsby-source-filesystem/src/create-remote-file-node.js]: This is looking good to me! The basic idea is that it's going to make parsing the extension a little more robust, which is great.

How about some tests validating the approach? Also just to sanity check this, I [created a Runkit](https://runkit.com/dschau/5bad2e64a12ba30012a6adfa) to check this out!

Comment (pieh) [reply packages/gatsby-source-filesystem/src/create-remote-file-node.js]: let's move `path.parse(Url.parse(url).pathname).ext` to separate function, that we can export and unit test against couple of urls including `https://scontent.xx.fbcdn.net/v/t51.2885-15/42078503_294439751160571_1602896118583132160_n.jpg?_nc_cat=101&oh=e30490a47409051c45dc3daacf616bc0&oe=5C5EA8EB`

Comment (DSchau) [reply packages/gatsby-source-filesystem/src/create-remote-file-node.js]: @oorestisime this is actually what I was going to suggest. If we move to a separate file/function, we can add some unit tests sorta like what I've done in the Runkit!

Comment (oorestisime) [reply packages/gatsby-source-filesystem/src/create-remote-file-node.js]: yes sounds perfect. added this with latest commit. thanks

File: packages/gatsby-source-filesystem/src/__tests__/create-remote-file-node.js [hunk 2]
```diff
@@ -0,0 +1,14 @@
+const { getRemoteFileExtension } = require(`../utils`)
+
+describe(`create remote file node`, () => {
+  it(`can correctly retrieve files extensions`, () => {
+    expect([
```
[the remark this answers is not in the document]
Comment (oorestisime) [reply packages/gatsby-source-filesystem/src/__tests__/create-remote-file-node.js]: ty, didnt occur to me like that :)

**Comment by pieh** [comment]
I think this looks great - one NIT: please rename test file to `utils.js` as we are testing only utils there right now

**Comment by oorestisime** [comment]
Oups sorry! corrected

**Pull request merged**